packages/dcos-integration-test/extra/test_dcos_diagnostics.py

```diff
@@ -640,35 +536,129 @@ def _get_dcos_diagnostics_health(z: zipfile.ZipFile, item: str):
-        assert len(properties) == len(required_status_fields), 'response must have the following fields: {}'.format(
-            required_status_fields
-        )
-        for required_status_field in required_status_fields:
-            assert required_status_field in properties, 'property {} not found'.format(required_status_field)
+    return node_private_public_ip_map
+
+
+@retrying.retry(wait_fixed=2000, stop_max_delay=120000,
+                retry_on_result=lambda x: x is False)
+def wait_for_diagnostics_job(dcos_api_session, last_datapoint):
+    response = check_json(dcos_api_session.health.get('/report/diagnostics/status/all'))
+    # find if the job is still running
+    job_running = False
+    percent_done = 0
+    for _, attributes in response.items():
+        assert 'is_running' in attributes, '`is_running` field is missing in response'
+        assert 'job_progress_percentage' in attributes, '`job_progress_percentage` field is missing in response'
+
+        if attributes['is_running']:
+            percent_done = attributes['job_progress_percentage']
+            logging.info("Job is running. Progress: {}".format(percent_done))
+            job_running = True
+            break
+
+    # if we ran this bit previously compare the current datapoint with the one we saved
+    if last_datapoint['time'] and last_datapoint['value']:
+        if percent_done <= last_datapoint['value']:
+            assert (datetime.datetime.now() - last_datapoint['time']) < datetime.timedelta(seconds=15), (
+                "Job is not progressing"
+            )
+    last_datapoint['value'] = percent_done
+    last_datapoint['time'] = datetime.datetime.now()
+
+    return not job_running
+
+
+# sometimes it may take extra few seconds to list bundles after the job is finished.
+@retrying.retry(stop_max_delay=5000)
+def wait_for_diagnostics_list(dcos_api_session):
+    assert _get_bundle_list(dcos_api_session), 'get a list of bundles timeout'
+
+
+def validate_node(nodes):
+    assert isinstance(nodes, list), 'input argument must be a list'
+    assert len(nodes) > 0, 'input argument cannot be empty'
+    required_fields = ['host_ip', 'health', 'role']
+
+    for node in nodes:
+        assert len(node) == len(required_fields), 'node should have the following fields: {}. Actual: {}'.format(
+            ', '.join(required_fields), node)
+        for required_field in required_fields:
+            assert required_field in node, '{} must be in node. Actual: {}'.format(required_field, node)
+
+        # host_ip, health, role fields cannot be empty
+        assert node['health'] in [0, 1], 'health must be 0 or 1'
+        assert node['host_ip'], 'host_ip cannot be empty'
+        assert node['role'], 'role cannot be empty'
+
+
+def validate_units(units):
+    assert isinstance(units, list), 'input argument must be list'
+    assert len(units) > 0, 'input argument cannot be empty'
+
+    for unit in units:
+        validate_unit(unit)
```

Comment:
Here is the only change in code I made. Other changes are only done by moving functions down